### PR TITLE
FOGL-653 Implement shutdown method in Micro-Service Management api

### PR DESCRIPTION
**Pull request by praveen-garg**
- Removed keyboard interrupt binding based stop
- Allow shutdown via `http://localhost:<core port>/foglamp/service/shutdown` POST only;
- Added TODO for scheduler timeout error (we don't raise it as of now)

This PR also take care of:

- [x] FOGL-955,

- [x] FOGL-956 

- [ ] (partially fixed FOGL-954)

 > everything stops with shutdown call except north plugins tasks cleanup i.e. FOGL-820

Logs:

```
FogLAMP[1615] ERROR: server: foglamp.services.core.server: Unable to stop the scheduler#012Traceback (most recent call last):#012  File "/home/foglamp/foglamp/FogLAMP/python/foglamp/services/core/server.py", line 556, in _stop_scheduler#012    await cls.scheduler.stop()#012  File "/home/foglamp/foglamp/FogLAMP/python/foglamp/services/core/scheduler/scheduler.py", line 863, in stop#012    raise TimeoutError("Timeout Error: Could not stop scheduler")#012TimeoutError: Timeout Error: Could not stop scheduler
FogLAMP[1615] INFO: server: foglamp.services.core.server: Shutting down the Southbound service COAP ...
FogLAMP[1615] INFO: server: foglamp.services.core.server: Shutting down the Southbound service CC2650POLL ...
FogLAMP[1699] INFO: coap_listen: foglamp.plugins.south.coap_listen.coap_listen: COAP plugin shut down.
FogLAMP[1701] INFO: cc2650poll: foglamp.plugins.south.cc2650poll.cc2650poll: CC2650 poll plugin shut down.
FogLAMP[1615] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Unregistered service instance id=fa02f9b0-0fdc-45a0-aedd-3024b3e5dfcc: <COAP, type=Southbound, protocol=http, address=0.0.0.0, service port=39528, management port=39528, status=1>
FogLAMP[1615] INFO: server: foglamp.services.core.server: Successfully shut down the Southbound service COAP.
FogLAMP[1615] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Unregistered service instance id=15e0e320-1539-43b8-8e7b-e2610fdd25f0: <CC2650POLL, type=Southbound, protocol=http, address=0.0.0.0, service port=46160, management port=46160, status=1>
FogLAMP[1615] INFO: server: foglamp.services.core.server: Successfully shut down the Southbound service CC2650POLL.
FogLAMP[1615] INFO: server: foglamp.services.core.server: Services monitoring stopped.
FogLAMP[1615] INFO: server: foglamp.services.core.server: Rest server stopped.
FogLAMP[1615] INFO: server: foglamp.services.core.server: Shutting down the Storage service FogLAMP Storage ...
FogLAMP Storage[1629]: Storage service shutdown in progress.
FogLAMP[1615] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Unregistered service instance id=96fafb24-d15b-4ed6-ac92-e1d1e48b6a93: <FogLAMP Storage, type=Storage, protocol=http, address=localhost, service port=39336, management port=45304, status=1>
FogLAMP Storage[1629]: Unregistered service 96fafb24-d15b-4ed6-ac92-e1d1e48b6a93.
FogLAMP Storage[1629]: Storage service shut down.
FogLAMP[1615] INFO: server: foglamp.services.core.server: Successfully shut down the Storage service FogLAMP Storage.
FogLAMP[1615] INFO: server: foglamp.services.core.server: Stopping the FogLAMP Core event loop. Good Bye!
```

> then some sending process pending tasks log..

### CODE REVIEW

**Comment by amarendra-dianomic** [comment]
Tried to shutdown vide rest apihttp://0.0.0.0:33371/foglamp/service/shutdown. Took screenshots of foglamp processes before and after call. Seems that Scheduler shutdown is not being done properly. Sending processes instances are still not terminated. Here is the result:
foglamp1

Also, A prompt "Error:" appears after the server exited:
foglamp2

**Comment by praveen-garg** [comment]
Yes, scheduler stop seems an issue; we need to handle it better via scheduler stop.

File: python/foglamp/services/core/server.py [hunk 1]
```diff
@@ -590,13 +627,47 @@ def stop_storage(cls):
...
+        :return: JSON payload with message key
+        :Example:
+            curl -X POST http://localhost:<core mgt port>/foglamp/service/shutdown
+        """
+        try:
+            # There are 3 types of services
+            #   - Core
+            #   - Storage
+            #   - Southbound
+            try:
```
Comment (praveen-garg) [inline python/foglamp/services/core/server.py]: I think we should not add this block, as these are started by scheduler hence scheduler should stop / shutdown these. 

Or
I should get all (not only southbound) and  
```
if svc._name == "FogLAMP Storage" or svc._name == "FogLAMP Core": 
    continue
else shutdown
```

Comment (MarkRiddoch) [reply python/foglamp/services/core/server.py]: If this is shutdown all of FogLAMP, then it should

disable new tasks from running
 shut down the south microservices first
Then all tasks and south microservices are down then shutdown storage and finally exit

File: python/foglamp/services/core/server.py [hunk 2]
```diff
@@ -257,7 +263,7 @@ def __start_storage(host, m_port):
         try:
             cmd_with_args = ['./services/storage', '--address={}'.format(host),
                              '--port={}'.format(m_port)]
-            subprocess.call(cmd_with_args, cwd=_SCRIPTS_DIR)
+            subprocess.call(cmd_with_args, cwd=_FOGLAMP_ROOT)
```
[the remark this answers is not in the document]
Comment (praveen-garg) [reply python/foglamp/services/core/server.py]: Okay, I will remove start storage call

File: python/foglamp/services/core/server.py [hunk 3]
```diff
@@ -404,56 +383,111 @@ def _register_core(cls, host, mgt_port, service_port):
...
+    #     await cls.core_app.shutdown()
+    #     await cls.core_server_handler.shutdown(60.0)
+    #     await cls.core_app.cleanup()
+    #     _logger.info("Core server stopped.")
+
+    @classmethod
+    async def stop_storage(cls):
+        """Stops Storage service """
+
+        # Or just do?!
```
[the remark this answers is not in the document]
Comment (praveen-garg) [reply python/foglamp/services/core/server.py]: Okay, I will remove these (ask) comments, thanks for the clarification

File: python/foglamp/services/core/server.py (hunk 1 above)
Comment (praveen-garg) [reply python/foglamp/services/core/server.py]: I believe scheduler should stop all these tasks and plugins and here we should call only stop scheduler which we are already doing here

File: python/foglamp/services/core/server.py [hunk 4]
```diff
@@ -672,8 +747,6 @@ def main():
         if command == 'start':
             Server().start()
         elif command == 'stop':
-            Server().stop_storage()
-            Server().stop_service_monitor()
-            # scheduler has signal binding
+            Server().stop()
```
Comment (praveen-garg) [inline python/foglamp/services/core/server.py]: I will remove `def main():` and possibly the entry point from setup.py as well ; as we are not using this; instead we are calling everything from shell scripts with `python3 -m ....`

**Comment by praveen-garg** [comment]
@amarendra-dianomic It should be unblocked to use that 455 luxury stuff

**Comment by amarendra-dianomic** [comment]
As PR#455 has been withdrawn,  ping and shutdown methods may need to be implemented independently in this PR.

**Pull request merged**